File: src/jev_factorio/craft_jobs.py
"""Strict, observation-only evidence for one native handcraft job.

The engine event receipt, not an inventory delta or a success message, identifies
production. No function here contacts the game or authorizes replay.
"""
from __future__ import annotations

from copy import deepcopy
from dataclasses import asdict, dataclass


class InvalidCraftEvidence(ValueError):
    """Missing, inconsistent or regressed native evidence."""


def natural(value: object, *, positive: bool = False) -> int:
    if type(value) is not int or value < int(positive):
        raise InvalidCraftEvidence("Invalid crafting counter")
    return value


def identifier(value: object) -> str:
    if not isinstance(value, str) or not value or len(value) > 128:
        raise InvalidCraftEvidence("Invalid crafting identity")
    return value


def counts(value: object, *, positive: bool = False) -> dict[str, int]:
    if not isinstance(value, dict) or not value or len(value) > 32:
        raise InvalidCraftEvidence("Invalid crafting item map")
    return {identifier(key): natural(amount, positive=positive) for key, amount in value.items()}


def receipt_for(parameters: dict, snapshot) -> dict:
    """Validate receipt shape and bind it to the current actor/session/request."""
    factory = snapshot.factory
    actor, receipt = factory.get("craft_job_actor"), factory.get("craft_job")
    if not isinstance(actor, dict) or not isinstance(receipt, dict):
        raise InvalidCraftEvidence("Native crafting evidence is unavailable")
    if (type(factory.get("craft_jobs_protocol")) is not int
            or factory.get("craft_jobs_protocol") != 1
            or receipt.get("paid") is not True
            or factory.get("player_connected") is not True
            or factory.get("player_bound") is not True
            or receipt.get("id") != parameters.get("receipt")
            or receipt.get("recipe") != parameters.get("recipe")
            or receipt.get("requested") != parameters.get("batches")
            or receipt.get("session_id") != snapshot.session_id
            or actor.get("session_id") != snapshot.session_id):
        raise InvalidCraftEvidence("Native crafting identity mismatch")
    for key in ("player_index", "unit_number", "surface_index", "force_index"):
        if natural(receipt.get(key), positive=True) != natural(actor.get(key), positive=True):
            raise InvalidCraftEvidence("Native crafting actor changed")
    identifier(receipt.get("id"))
    identifier(receipt.get("recipe"))
    requested = natural(receipt.get("requested"), positive=True)
    if not requested <= 200 or natural(receipt.get("accepted")) != requested:
        raise InvalidCraftEvidence("Native crafting request was not fully accepted")
    finished = natural(receipt.get("finished"))
    started, last = natural(receipt.get("started_tick")), natural(receipt.get("last_progress_tick"))
    if finished > requested or not started <= last <= snapshot.tick:
        raise InvalidCraftEvidence("Native crafting progress is inconsistent")
    inputs = counts(receipt.get("inputs"), positive=True)
    baseline = counts(receipt.get("baseline"))
    outputs = counts(receipt.get("outputs"), positive=True)
    if (len(outputs) != 1 or set(outputs) != set(baseline) or set(inputs) & set(outputs)
            or any(total % requested for total in outputs.values())):
        raise InvalidCraftEvidence("Unsupported native crafting products")
    if receipt.get("status") not in {"running", "completed"}:
        raise InvalidCraftEvidence("Native crafting was cancelled or contaminated")
    if receipt["status"] == "completed":
        completed = natural(receipt.get("completed_tick"))
        if finished != requested or not last <= completed <= snapshot.tick:
            raise InvalidCraftEvidence("Native crafting completion is inconsistent")
    elif finished >= requested:
        raise InvalidCraftEvidence("Native crafting status is inconsistent")
    if receipt.get("queue_valid") is not True:
        raise InvalidCraftEvidence("Native crafting queue changed")
    return receipt


def craft_complete(parameters: dict, snapshot) -> bool:
    try:
        receipt = receipt_for(parameters, snapshot)
        return receipt["status"] == "completed" and all(
            snapshot.inventory.get(item, 0) >= receipt["baseline"][item] + amount
            for item, amount in receipt["outputs"].items()
        )
    except (InvalidCraftEvidence, TypeError, AttributeError):
        return False


@dataclass
class CraftJob:
    """A positively acknowledged job; its outputs stay locked until verified."""

    parameters: dict
    plan_id: str
    goal: str
    session_id: str
    actor: dict
    inputs: dict[str, int]
    outputs: dict[str, int]
    baseline: dict[str, int]
    started_tick: int
    deadline_tick: int
    finished: int = 0
    last_progress_tick: int = 0
    failed: str = ""

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> CraftJob:
        try:
            job = cls(**deepcopy(data))
            if (not isinstance(job.parameters, dict)
                    or set(job.parameters) != {"recipe", "batches", "receipt"}):
                raise InvalidCraftEvidence("Invalid background request")
            for key in ("recipe", "receipt"):
                identifier(job.parameters[key])
            requested = natural(job.parameters["batches"], positive=True)
            if requested > 200:
                raise InvalidCraftEvidence("Background batch exceeds budget")
            for value in (job.plan_id, job.goal, job.session_id):
                identifier(value)
            if (not isinstance(job.actor, dict)
                    or set(job.actor) != {"player_index", "unit_number", "surface_index", "force_index"}):
                raise InvalidCraftEvidence("Invalid background actor")
            for value in job.actor.values():
                natural(value, positive=True)
            counts(job.inputs, positive=True)
            counts(job.outputs, positive=True)
            counts(job.baseline)
            if (len(job.outputs) != 1 or set(job.outputs) != set(job.baseline)
                    or set(job.inputs) & set(job.outputs)
                    or any(total % requested for total in job.outputs.values())):
                raise InvalidCraftEvidence("Invalid background products")
            if not (natural(job.started_tick) <= natural(job.last_progress_tick)
                    < natural(job.deadline_tick)):
                raise InvalidCraftEvidence("Invalid background deadline")
            if natural(job.finished) >= requested or not isinstance(job.failed, str) or len(job.failed) > 160:
                raise InvalidCraftEvidence("Invalid background progress")
            return job
        except (TypeError, KeyError, AttributeError) as error:
            raise InvalidCraftEvidence("Invalid background checkpoint") from error

    @classmethod
    def admit(cls, plan, pending: dict, snapshot, catalog) -> CraftJob:
        step = plan.steps[0]
        if (len(plan.steps) != 1 or step.action != "factory_craft_job"
                or pending.get("dispatch") != "returned" or pending.get("action") != step.action):
            raise InvalidCraftEvidence("Only acknowledged single-step crafts may run in background")
        receipt = receipt_for(step.parameters, snapshot)
        recipe = catalog.recipes.get(step.parameters["recipe"], {})
        batches = step.parameters["batches"]
        expected_inputs = {entry["name"]: entry["amount"] * batches
                           for entry in recipe.get("ingredients", []) if entry["type"] == "item"}
        expected_outputs = {entry["name"]: entry["amount"] * batches
                            for entry in recipe.get("products", []) if entry["type"] == "item"}
        if (receipt["inputs"] != expected_inputs or receipt["inputs"] != step.costs
                or receipt["outputs"] != expected_outputs
                or receipt["status"] != "running"
                or receipt["started_tick"] < pending["started_tick"]):
            raise InvalidCraftEvidence("Native craft does not match the committed plan")
        data = dict(
            parameters=deepcopy(step.parameters), plan_id=plan.id, goal=plan.goal,
            session_id=snapshot.session_id,
            actor={key: receipt[key] for key in ("player_index", "unit_number", "surface_index", "force_index")},
            inputs=deepcopy(receipt["inputs"]), outputs=deepcopy(receipt["outputs"]),
            baseline=deepcopy(receipt["baseline"]), started_tick=receipt["started_tick"],
            # Write-ahead persistence can delay dispatch after its observation.
            # Only this validated, acknowledged native start anchors execution;
            # the resulting deadline stays fixed across polls and resumes.
            deadline_tick=receipt["started_tick"] + step.timeout_ticks,
            finished=receipt["finished"], last_progress_tick=receipt["last_progress_tick"],
        )
        return cls.from_dict(data)

    def observe(self, snapshot) -> bool:
        """Update monotonic evidence; return True only for verified completion."""
        if self.failed:
            raise InvalidCraftEvidence(self.failed)
        receipt = receipt_for(self.parameters, snapshot)
        if (snapshot.session_id != self.session_id
                or any(receipt[key] != value for key, value in self.actor.items())
                or any(receipt[key] != getattr(self, key) for key in ("inputs", "outputs", "baseline", "started_tick"))
                or receipt["finished"] < self.finished
                or receipt["last_progress_tick"] < self.last_progress_tick):
            raise InvalidCraftEvidence("Background evidence changed or regressed")
        if receipt["status"] == "completed":
            if receipt["completed_tick"] > self.deadline_tick or not craft_complete(self.parameters, snapshot):
                raise InvalidCraftEvidence("Background output is late or missing")
            return True
        if snapshot.tick >= self.deadline_tick:
            raise InvalidCraftEvidence("Background crafting exceeded its deadline")
        # The event fires before insertion. Routine observations happen later;
        # a shortfall here is external output consumption or missing evidence.
        for item, total in self.outputs.items():
            earned = receipt["finished"] * (total // self.parameters["batches"])
            if snapshot.inventory.get(item, 0) < self.baseline[item] + earned:
                raise InvalidCraftEvidence("Background output was consumed before verification")
        self.finished, self.last_progress_tick = receipt["finished"], receipt["last_progress_tick"]
        return False

    def permits(self, step) -> bool:
        """No second craft, construction, or consumption of any locked output."""
        if self.failed or step.action not in {"factory_gather", "factory_insert", "factory_extract", "factory_wait"}:
            return False
        parameters = step.parameters or {}
        touched = set(step.costs or {})
        touched.update(value for value in (parameters.get("item"), parameters.get("resource")) if value)
        return not (touched & self.outputs.keys())

File: src/jev_factorio/lua/craft_jobs.lua
-- Opt-in receipt tracking. Never awards items, changes speed, or replays work.
local campaign = assert(storage.campaign)
campaign.craft_jobs = campaign.craft_jobs or {}
local jobs = campaign.craft_jobs
local buffers, routes = storage.output_buffers, storage.input_routes
local observer = campaign.observe
if routes and observer == routes.observer then observer = routes.previous_observe end
if buffers and observer == buffers.observer then observer = buffers.previous_observe end
if jobs.observe_wrapper and observer == jobs.observe_wrapper
    and campaign.observe ~= jobs.observe_wrapper then
    for _, entry in ipairs({
        {defines.events.on_pre_player_crafted_item, jobs.pre_handler},
        {defines.events.on_player_cancelled_crafting, jobs.cancel_handler},
        {defines.events.on_player_crafted_item, jobs.crafted_handler}
    }) do
        assert(entry[2] and script.get_event_handler(entry[1]) == entry[2],
            "Craft event handler changed")
    end
    return
end

local function actor()
    local player = storage.fair.actor()
    assert(type(storage.jev_session_id) == "string", "Missing native session")
    return player, {
        session_id = storage.jev_session_id, player_index = player.index,
        unit_number = player.character.unit_number,
        surface_index = player.surface.index, force_index = player.force.index
    }
end

local function invalidate(reason)
    if jobs.job then jobs.job.status, jobs.job.error = "invalid", reason end
end

local function same_actor(job, identity)
    for _, key in ipairs({"session_id", "player_index", "unit_number", "surface_index", "force_index"}) do
        if job[key] ~= identity[key] then return false end
    end
    return true
end

local function queue_valid(player, job)
    local remaining = 0
    for _, entry in pairs(player.crafting_queue or {}) do
        if entry.recipe ~= job.recipe or entry.prerequisite then return false end
        remaining = remaining + entry.count
    end
    return remaining == job.requested - job.finished
end

campaign.begin_craft_job = function(id, recipe_name, batches)
    local player, identity = actor()
    assert(type(id) == "string" and #id > 0 and #id <= 128, "Invalid job identity")
    assert(type(batches) == "number" and batches % 1 == 0 and batches >= 1 and batches <= 200,
        "Invalid craft batch")
    assert(not jobs.job or (jobs.job.status == "completed" and jobs.job.id ~= id),
        "Previous craft is unresolved or request was already used")
    assert(player.crafting_queue_size == 0, "Handcraft queue must be empty")
    local recipe = player.force.recipes[recipe_name]
    assert(recipe and recipe.enabled and #recipe.products == 1, "Unsupported native recipe")
    local product = recipe.products[1]
    assert(product.type == "item" and (not product.probability or product.probability == 1)
        and product.amount and product.amount > 0 and product.amount % 1 == 0,
        "Background craft needs one deterministic item product")
    local inputs, before = {}, {}
    assert(#recipe.ingredients > 0, "Background craft needs paid ingredients")
    for _, ingredient in pairs(recipe.ingredients) do
        assert(ingredient.type == "item" and ingredient.name ~= product.name
            and ingredient.amount > 0 and ingredient.amount % 1 == 0, "Unsupported craft ingredient")
        inputs[ingredient.name] = (inputs[ingredient.name] or 0) + ingredient.amount * batches
    end
    for name, amount in pairs(inputs) do
        before[name] = player.get_item_count(name)
        assert(before[name] >= amount, "All immediate ingredients must be carried")
    end
    local job = {
        id = id, recipe = recipe_name, requested = batches, accepted = 0,
        finished = 0, started_tick = game.tick, last_progress_tick = game.tick,
        baseline = {[product.name] = player.get_item_count(product.name)},
        outputs = {[product.name] = product.amount * batches}, inputs = inputs,
        status = "preparing", queue_valid = false, paid = false
    }
    for key, value in pairs(identity) do job[key] = value end
    jobs.job = job -- intent survives a command failure; never silently retry it
    jobs.submitting = true
    local ok, accepted = pcall(function()
        return player.begin_crafting{count = batches, recipe = recipe_name, silent = true}
    end)
    jobs.submitting = false
    if not ok then invalidate("enqueue_error"); error("Native crafting enqueue failed") end
    job.accepted = accepted
    if accepted ~= batches then invalidate("partial_acceptance"); error("Partial native craft acceptance") end
    for name, amount in pairs(inputs) do
        if before[name] - player.get_item_count(name) ~= amount then
            invalidate("input_debit_mismatch"); error("Native craft input debit mismatch")
        end
    end
    if job.status ~= "preparing" then error("Craft changed while submitting") end
    job.paid, job.status = true, "running"
    job.queue_valid = queue_valid(player, job)
    if not job.queue_valid then invalidate("queue_mismatch"); error("Native craft queue mismatch") end
    if campaign.successor_craft_paid then campaign.successor_craft_paid(job,before) end
    rcon.print("Native craft accepted; output remains unverified")
end

local function install(event_id, key, callback)
    local current = script.get_event_handler(event_id)
    assert(not jobs[key] or current == jobs[key],
        "Craft event handler changed; reconcile before attachment")
    if current ~= jobs[key] then jobs[key .. "_previous"] = current end
    local previous = jobs[key .. "_previous"]
    jobs[key] = function(event)
        if previous then previous(event) end
        callback(event)
    end
    script.on_event(event_id, jobs[key])
end

install(defines.events.on_pre_player_crafted_item, "pre_handler", function(event)
    local job = jobs.job
    if job and job.status == "running" and event.player_index == job.player_index and not jobs.submitting then
        invalidate("external_queue_change")
    end
end)
install(defines.events.on_player_cancelled_crafting, "cancel_handler", function(event)
    local job = jobs.job
    if job and job.status == "running" and event.player_index == job.player_index then
        invalidate("craft_cancelled")
    end
end)
install(defines.events.on_player_crafted_item, "crafted_handler", function(event)
    local job = jobs.job
    if not job or job.status ~= "running" or event.player_index ~= job.player_index then return end
    local ok, player, identity = pcall(actor)
    if not ok or not same_actor(job, identity) then invalidate("actor_changed"); return end
    local stack = event.item_stack
    local expected = stack and stack.valid_for_read and job.outputs[stack.name]
    if not event.recipe or event.recipe.name ~= job.recipe or not expected
        or stack.count ~= expected / job.requested
        or (stack.quality and stack.quality.name ~= "normal") then
        invalidate("unexpected_craft_event"); return
    end
    job.finished, job.last_progress_tick = job.finished + 1, game.tick
    if job.finished == job.requested then
        job.status, job.completed_tick = "completed", game.tick
    elseif job.finished > job.requested then invalidate("excess_craft_events") end
end)

if campaign.observe ~= jobs.observe_wrapper then jobs.previous_observe = campaign.observe end
local previous_observe = jobs.previous_observe
jobs.observe_wrapper = function()
    local result = previous_observe()
    local player, identity = actor()
    result.craft_jobs_protocol, result.craft_job_actor = 1, identity
    -- FLE's earlier inventory read can precede a crafting event. Capture the
    -- main inventory in this same Lua observation as the receipt and game tick.
    local inventory = {}
    for _, stack in pairs(player.get_main_inventory().get_contents()) do
        inventory[stack.name] = (inventory[stack.name] or 0) + stack.count
    end
    result.craft_job_inventory = {tick = game.tick, items = inventory}
    local job = jobs.job
    if job then
        if not same_actor(job, identity) then invalidate("actor_changed") end
        job.queue_valid = queue_valid(player, job)
        if not job.queue_valid and job.status ~= "preparing" then invalidate("queue_mismatch") end
        result.craft_job = job
    end
    return result
end
campaign.observe = jobs.observe_wrapper
